src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java

```diff
@@ -4230,6 +4234,87 @@ public void onOpenProfileEvent(OpenProfileEvent event)
       onShowProfiler(event);
    }
    
+   @Override
+   public void onRequestDocumentSave(RequestDocumentSaveEvent event)
+   {
+      final JsArrayString ids = event.getData().getDocumentIds();
+      
+      // we use a timer that fires the document save completed event,
+      // just to ensure the server receives a response even if something
+      // goes wrong during save or some client-side code throws. unfortunately
+      // the current save code is wired up in such a way that it's difficult
+      // to distinguish a success from failure, so we just make sure the
+      // server receives a response after 5s (defaulting to failure)
+      final Mutable<Boolean> savedSuccessfully = new Mutable<Boolean>(false);
+      final Timer completedTimer = new Timer()
+      {
+         @Override
+         public void run()
+         {
+            server_.requestDocumentSaveCompleted(
```

Comment:
How does this work with source windows? Seems like as written each window will wind up calling this, which is probably not what you want. If it's not clear to you how to wire this up ping me (the problem is already largely solved in the existing quit logic).